domains/client/domain-executor/src/domain_block_processor.rs

```diff
@@ -182,6 +196,130 @@ where
         }
     }
 
+    pub(crate) fn fork_aware_block_imports(
+        &self,
+        primary_hash: PBlock::Hash,
+        primary_number: NumberFor<PBlock>,
+    ) -> sp_blockchain::Result<ForkAwareBlockImports<Block, PBlock>> {
+        if primary_number == One::one() {
+            return Ok(ForkAwareBlockImports {
+                initial_parent: (self.client.info().genesis_hash, Zero::zero()),
+                primary_imports: vec![HashAndNumber {
+                    hash: primary_hash,
+                    number: primary_number,
+                }],
+            });
+        }
+
+        let best_hash = self.client.info().best_hash;
+        let best_number = self.client.info().best_number;
+        let best_receipt = crate::aux_schema::load_execution_receipt::<
+            _,
+            Block::Hash,
+            NumberFor<PBlock>,
+            PBlock::Hash,
+        >(&*self.client, best_hash)?
+        .ok_or_else(|| {
+            sp_blockchain::Error::Backend(format!("Execution receipt not found for {best_hash:?}"))
+        })?;
+
+        let primary_from = best_receipt.primary_hash;
+        let primary_to = primary_hash;
+
+        if primary_from == primary_to {
+            return Err(sp_blockchain::Error::Application(Box::from(
+                "Primary block {primary_hash:?} has already been processed.",
+            )));
+        }
+
+        let route =
+            sp_blockchain::tree_route(&*self.primary_chain_client, primary_from, primary_to)?;
+
+        let retracted = route.retracted();
+        let enacted = route.enacted();
+
+        tracing::trace!(
+            ?retracted,
+            ?enacted,
+            common_block = ?route.common_block(),
+            "Calculating ForkAwareBlockImports on #{best_number},{best_hash:?}"
+        );
+
+        match (retracted.is_empty(), enacted.is_empty()) {
+            (true, false) => {
```

Comment:
With below comments, all branches without `unreachable` when corrected, will end up being exactly the same (`common_block_number` thing will have to be done in all cases instead of `.info()`) unless I read it incorrectly. Which means we probably don't need `match` here at all.

Reply:
@liuchengxu so with refactoring that you did we have 5 branches, one of which returns `None` (though could have returned an empty vector of pending primary block imports with the same result), two are unreachable and two remaining only differ in `initial_parent` construction. Moreover, I think second `initial_parent` construction is a super-set of the first one. What about removing `match` now and just use the last version as a universal return value for much simplified code?

Reply:
Well, the only thing we care here is what blocks from primary chain were not yet processed on a domain relatively to some parent. The rest IMHO is an implementation detail, so I don't see how `Option<Vec<T>>` is helping here.